src/rules/checkTagNames.js

```diff
@@ -54,12 +110,83 @@ export default iterateJsdoc(({
       });
   }
 
+  const isInAmbientContext = (subNode) => {
+    return subNode.type === 'Program' ?
+      context.getFilename().endsWith('.d.ts') :
+      Boolean(subNode.declare) || isInAmbientContext(subNode.parent);
+  };
+
+  const tagIsRedundantWhenTyped = (jsdocTag) => {
+    if (!typedTagsUnnecessaryOutsideDeclare.has(jsdocTag.tag)) {
+      return false;
+    }
+
+    if (jsdocTag.description.trim() === 'default' || jsdocTag.name === 'default') {
+      return false;
+    }
+
+    if (context.getFilename().endsWith('.d.ts') && node.parent.type === 'Program') {
+      return false;
+    }
+
+    if (isInAmbientContext(node)) {
+      return false;
+    }
+
+    return true;
+  };
+
+  const tagRedundantlyChecksParamOrReturnTyped = (jsdocTag) => {
+    if (!typedTagsParamReturn.has(jsdocTag.tag)) {
+      return false;
+    }
+
+    if (!jsdocTag.description) {
+      report(`'@${jsdocTag.tag}' without a description is redundant when using a type system.`, null, jsdocTag);
```

Comment:
In the case of `report`, the second argument should be a fixer function. It normally is used with a fixer argument which can then have its `replaceText` method called to perform string manipulations.

However, I would instead recommend using `utils.reportJSDoc()` which allows one to manipulate the AST, and have it be stringified back.

You can remove a tag by its index with `utils.removeTag()`.

Note also that in some cases, you may need to use the `specRewire` option of `reportJSDoc` if rewriting the tags or otherwise they won't be reflected in the stringification.

Reply:
Thanks for the info. I'm still stuck though 😞 sorry.

d624e9f adds `utils.reportJSDoc` with `specRewire=true` and `utils.removeTag(tagIndex)` usage. But the `output` in the spec is still not being updated:

```plaintext
 1) check-tag-names
       invalid
         
        /** @type {string} */
        let a;
      :

      AssertionError [ERR_ASSERTION]: Output is incorrect.
      + expected - actual

       
      -        /** @type {string} */
               let a;
             
```

Am I missing something still?

Reply:
There was an issue with rewiring, but I wouldn't think that should apply here. Could you add this commit to a branch to let me work with it more easily? It might have to do with a bug in `removeTag` in regards to 0-line tags

Reply:
I'm guessing that we have two problems: one with removing a tag on what we call the 0th line, and two the desire to have a utility to remove the entire block if it is empty (since the utility to remove a tag should really just remove the tag and leave the block in place, even if that means leaving just `/** */`).

Reply:
I've pushed two commits, one of which fixes (and the other is an unrelated nit I wanted to fix in testing)

Reply:
Awesome, thanks! I'll fill out the rest of the fixers and mark this as ready once that's done.